### ISSUE TITLE
PY_SSIZE_T_CLEAN conflicts with Py_LIMITED_API

### ISSUE THREAD

**Issue by ce7451b2-bb9b-4ec5-95f7-6b62a7d6240c** [python/cpython#71686]
BPO | [27499](https://bugs.python.org/issue27499)
--- | :---
Nosy | @vstinner, @dholth, @serhiy-storchaka

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2016-07-12.15:59:47.480>
labels = ['extension-modules']
title = 'PY_SSIZE_T_CLEAN conflicts with Py_LIMITED_API'
updated_at = <Date 2020-06-10.16:48:48.291>
user = 'https://github.com/dholth'
```

bugs.python.org fields:
```python
activity = <Date 2020-06-10.16:48:48.291>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2016-07-12.15:59:47.480>
creator = 'dholth'
dependencies = []
files = []
hgrepos = []
issue_num = 27499
keywords = []
message_count = 6.0
messages = ['270252', '270255', '270256', '270260', '371221', '371222']
nosy_count = 3.0
nosy_names = ['vstinner', 'dholth', 'serhiy.storchaka']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue27499'
versions = ['Python 3.5']
```

</p></details>

**Comment by sahlberg** [comment]
SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats

This change hurt me too. A python c module I depend on just started doing this when I upgraded to Fedora36.
So, how do I fix this so it works again like in python3.8 ?

**Comment by vstinner** [comment]
> So, how do I fix this so it works again like in python3.8 ?

Modify the extension to use Py_ssize_t and define PY_SSIZE_T_CLEAN macro.

**Comment by vstinner** [comment]
Python 3.10 change (issue #85115):

> The PY_SSIZE_T_CLEAN macro must now be defined to use PyArg_ParseTuple() and Py_BuildValue() formats which use #: es#, et#, s#, u#, y#, z#, U# and Z#. See Parsing arguments and building values and PEP 353. (Contributed by Victor Stinner in bpo-40943.)

**Comment by vstinner** [comment]
@encukou: You may be interested by this tricky issue :-) I don't know its status in Python 3.12.

**Comment by encukou** [comment]
With `PY_SSIZE_T_CLEAN`, the `PyArg_ParseTuple` name is redefined as `_PyArg_ParseTuple_SizeT`, which passes an extra flag to the internal routines. Handling the affected format specifiers fails without the flag.
Both `_PyArg_ParseTuple_SizeT` and `PyArg_ParseTuple` are present in the ABI since 3.3. So, yes, 3.2 was buggy – use `0x03030000`.

The situation is unfortunate for non-C languages (which don't see macros).

Of course, removing the functionality was disruptive, like any removal of functionality. It wouldn't go away if I'd been in charge.
(It's not an ABI break – the module can still be loaded, it raises an exception at runtime – but that's little consolation to the affected user.)
Define `PY_SSIZE_T_CLEAN`, define `Py_LIMITED_API` as `0x3yy0000` (yy>03), recompile, and the module should again work on Python 3.yy and above.

**Comment by vstinner** [comment]
If recent versions of Python are fine, can we just close the issue?

**Comment by serhiy-storchaka** [comment]
How non-C languages work with variable-argument C functions? How they work with pointers to ssize_t?

**Comment by encukou** [comment]
> If recent versions of Python are fine, can we just close the issue?

Defining `Py_LIMITED_API` to `3` is still common, and IMO we can do better there. But it's a low priority for me :(

**Comment by erlend-aasland** [comment]
Is this issue still relevant after #106315? AFAICS, we can close this as outdated, no?

**Comment by vstinner** [comment]
> Is this issue still relevant after https://github.com/python/cpython/pull/106315? AFAICS, we can close this as outdated, no?

I close the issue. While the history is complicated, the root stable ABI issue has been fixed in Python 3.13 by commit adccff3b3f9fbdb58cb4b8fde92456e6dd078af0 of gh-104922.

History:

* Python 3.8: Use of # variants of formats in parsing or building value (e.g. PyArg_ParseTuple(), Py_BuildValue(), PyObject_CallFunction(), etc.) without PY_SSIZE_T_CLEAN defined raises DeprecationWarning
* Python 3.10: The PY_SSIZE_T_CLEAN macro must now be defined to use PyArg_ParseTuple() and Py_BuildValue() formats which use #: es#, et#, s#, u#, y#, z#, U# and Z#.
* Python 3.13: You no longer have to define the PY_SSIZE_T_CLEAN macro before including Python.h when using # formats in format codes. APIs accepting the format codes always use Py_ssize_t for # formats.

At the ABI level, Python <= 3.12 calls different functions if PY_SSIZE_T_CLEAN is defined. Example:

```c
#ifdef PY_SSIZE_T_CLEAN
#define PyArg_Parse                     _PyArg_Parse_SizeT
...
```

Python >= 3.13 calls the same function if PY_SSIZE_T_CLEAN macro is defined or not. The macro PY_SSIZE_T_CLEAN is ignored by Python 3.13 and newer.

If you consider Python >= 3.13, the stable ABI is fixed.

Python 3.11 and 3.12 still have the ABI issue. From what I understood, PY_SSIZE_T_CLEAN was never compatible with the stable ABI.

Python 3.10 and newer always use `Py_ssize_t` for size arguments, instead of `int`.

**Comment by mxmlnkn** [comment]
Now I am stumbling into [this](https://discuss.python.org/t/c-stable-api-py-ssize-t-clean-error/80159/1) issue. I.e. my code compiled with Python 3.13 with the limited API to get an abi3, which passes [abi3audit](https://github.com/pypa/abi3audit), does not work on Python 3.9 because I get:

> SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats

even though I did define that macro, but it seems to be ignored now.

Is [this](https://discuss.python.org/t/c-stable-api-py-ssize-t-clean-error/80159/4) answer the correct way to deal with this?

Full reproducer using Github Actions is [in this branch](https://github.com/mxmlnkn/fast-zip-decryption/tree/py_ssize_t-issue).

**Comment by vstinner** [comment]
Can you replace `"y#"` format (Py_ssize_t) with `"y*"` format (Py_buffer)?

**Comment by cakemanny** [comment]
`"y*"` is not available in the stable API until Python 3.11 ... so I don't think that's an option. (At least it wasn't for me).

**Comment by vstinner** [comment]
Oh right, sorry, I forgot that Py_buffer was only added recently (3.11) to the limited C API.

**Comment by vstinner** [comment]
> Now I am stumbling into [this](https://discuss.python.org/t/c-stable-api-py-ssize-t-clean-error/80159/1) issue. I.e. my code compiled with Python 3.13 with the limited API to get an abi3, which passes [abi3audit](https://github.com/pypa/abi3audit), does not work on Python 3.9

Well, one pratical solution is to build the C extension with Python 3.9.

**Comment by mxmlnkn** [comment]
Building with Python 3.12 also works fine. So, I guess I can live with that until 3.12 becomes EOL and slowly disappears from Github Actions end of 2028. And it almost makes sense, as "building on the oldest supported systems" is the standard advice for manylinux wheels.

But, it is unfortunate that building an abi3 wheel is not arbitrarily backwards and forwards compatible any longer with this change. I can imagine that I am not the only one accidentally creating broken abi3 wheels. I am just glad that I tested the abi3 wheel with all Python versions in my CI because I do not fully understand the limited C API / stable ABI and do not fully trust abi3audit.

**Comment by mxmlnkn** [comment]
Maybe something like this anywhere in Python.h 3.13+ could prevent such problems?

```c++
#if defined(Py_LIMITED_API) && (Py_LIMITED_API+0 <= 0x030120000) && defined(PY_SSIZE_T_CLEAN)
    #error "... recommended workaround ..."
#endif
```

**Comment by zooba** [comment]
> it is unfortunate that building an abi3 wheel is not arbitrarily backwards and forwards compatible any longer with this change

There have been plenty of other changes to make this true - I think 3.6 might've been the last version that had full forwards/backwards compatibility?

You "just" need to compile with the _earliest_ version that you want to support. Practically everything (not quite, but nearly) is forward compatible.

**Comment by mxmlnkn** [comment]
Good to know. Should this be reflected in the documentation?

> Define Py_LIMITED_API to the value of [PY_VERSION_HEX](https://docs.python.org/3/c-api/apiabiversion.html#c.PY_VERSION_HEX) corresponding to the lowest Python version your extension supports. The extension will be ABI-compatible with all Python 3 releases from the specified one onward, and can use Limited API introduced up to that version.

The current documentation suggests to me that backward compatibility is/was a goal at some point. Ok, reading the fine print, it is **ABI**-backward-compatible, just not "runtime"-backward-compatible... But who would want that?

**Comment by zooba** [comment]
Last time I looked, it was specified in the documentation well enough. I forget exactly where.

And yes, it _was_ meant to be forward and backward compatible, but it turns out that is impossible to maintain, and we've just gotten sloppier about what breakage is allowed over time.

That said, this particular case is pretty subtle, so not entirely surprised it was missed. And for the most part we've been good about not _changing_ APIs that already existed, so if you stick to an old set of APIs then you'll mostly be fine even with a newer set of headers (though the easiest way to validate this is to use the older headers, so that's the usual suggestion).
